README.md
# Exercise: Logging Commerce Events

This repo contains a simple commerce site to practice implementing commerce events. The application provides a `Run scenario` button which when clicked will run a test scenario (described below) and give you feedback on the implemnted events.

## The scenario

The test scenrio simulates a simple checkout flow. It will:

1. Add a random product to the cart.
2. Navigate to the checkout page.
3. Pay
4. Navigate to the search page.

When the scenario is complete, a modal will show the intercepted events and feedback on the payloads.

## The exercise

The exercise is to implement the following events:

1. A click event emitted when the `Add to cart` button is clicked.
2. An add-to-cart event emitted when the `Add to cart` button is clicked.
3. A page view event emitted when the checkout page is rendered.
4. A purchase event emitted when the `Pay` button on the checkout page is clicked.
5. A page view emitted when the search page is rendered.

Placeholder functions that need to be implemented are included in the code. Search for `logClick`, `logAddToCart`, `logPageView` and `logPurchase`.

Refer to the [documentation here](https://docs.coveo.com/en/3188/coveo-for-commerce/log-commerce-events) for information on how to emit the events and their expected payloads.

To test your implementation, click the `Run scenario` button.

## Scoring

Each implemented event provides a maximum of 2 points.

- 1 point for emitting the event.
- A maximum of 1 point if all keys in the payload are correct. e.g. if only 50% of the payload is valid, 0.5 points will be awarded.

## Setup

Pick one of the following setup options:

### In browser

- Code in the [online sandbox](https://codesandbox.io/p/github/samisayegh/commerce-store/main?file=/README.md).

### Local machine

- Clone the repo
- Install dependencies: `npm i`
- Start the server: `npm start`


server/CHANGELOG.md
## 1.35.22 (2023-03-27)


### Bug Fixes

* **deps:** update all dependencies  and jest snaps j:cdx-227 ([#1137](https://github.com/coveo/cli/issues/1137)) ([b72dc31](https://github.com/coveo/cli/commits/b72dc314043174ef9afaadb03e066c8830d7acc1)), closes [#1103](https://github.com/coveo/cli/issues/1103)
* **deps:** update all dependencies  and jest snaps j:cdx-227 ([#1149](https://github.com/coveo/cli/issues/1149)) ([2f2273c](https://github.com/coveo/cli/commits/2f2273c7d86f2a2a8414ebbdf8cddb800c888e96)), closes [#1144](https://github.com/coveo/cli/issues/1144)



## 1.35.20 (2023-03-01)

### Bug Fixes

- **deps:** update all dependencies j:cdx-227 ([#1090](https://github.com/coveo/cli/issues/1090)) ([ae24142](https://github.com/coveo/cli/commits/ae24142d3ae4b68e04951301eb157f9e7ddb1095))

## 1.35.19 (2023-01-12)

## 1.35.17 (2022-12-14)

### Bug Fixes

- **deps:** update all dependencies j:cdx-227 ([#1053](https://github.com/coveo/cli/issues/1053)) ([f218b3f](https://github.com/coveo/cli/commits/f218b3f2d8070890da1501622f34b757cd2ad3d9))
- **deps:** update all dependencies j:cdx-227 ([#1066](https://github.com/coveo/cli/issues/1066)) ([12533ea](https://github.com/coveo/cli/commits/12533ea7c8182d183a7158945de742202eac0b69))

## 1.35.16 (2022-11-29)

### Bug Fixes

- **deps:** update all dependencies j:cdx-227 ([#1032](https://github.com/coveo/cli/issues/1032)) ([e1ccdeb](https://github.com/coveo/cli/commits/e1ccdebcd1908b750df365961187013bc025ebdf))
- **deps:** update all dependencies j:cdx-227 ([#1036](https://github.com/coveo/cli/issues/1036)) ([1439a40](https://github.com/coveo/cli/commits/1439a406166f7342e1d91481e81de6c63a7b15f1))

## 1.35.15 (2022-11-15)

### Bug Fixes

- **deps:** update all dependencies j:cdx-227 ([#1015](https://github.com/coveo/cli/issues/1015)) ([0b6a70b](https://github.com/coveo/cli/commits/0b6a70b0c1539648ea3a24eefc8fb01b5b7173db))
- **deps:** update all dependencies j:cdx-227 ([#1022](https://github.com/coveo/cli/issues/1022)) ([e181d70](https://github.com/coveo/cli/commits/e181d700e19445bf912b3ecd370f7eabdb777d71))
- **deps:** update all dependencies j:cdx-227 ([#1026](https://github.com/coveo/cli/issues/1026)) ([c6ca8dd](https://github.com/coveo/cli/commits/c6ca8ddf8dde39eb69cd99da3e9d897db873e4c6))

## 1.35.14 (2022-10-13)

## 1.35.13 (2022-09-21)

## 1.35.10 (2022-09-07)

## 1.35.8 (2022-08-29)

## 1.34.2 (2022-08-17)

# 1.34.0 (2022-08-10)

# 1.33.0 (2022-07-26)

# 1.32.0 (2022-07-13)

### Bug Fixes

- **deps:** update all dependencies j:cdx-227 ([#854](https://github.com/coveo/cli/issues/854)) ([cd042cc](https://github.com/coveo/cli/commits/cd042cc9d08898a9cad39c644a99fa0dd64ba77a))

# 1.31.0 (2022-06-29)

### Features

- **snapshot:** add `deletionScope` on `/apply` ([#849](https://github.com/coveo/cli/issues/849)) ([7b7fcb2](https://github.com/coveo/cli/commits/7b7fcb2ed1d0fc73d0ee7500c91aed85fbfd0dc6))

## 1.29.1 (2022-06-03)

# 1.29.0 (2022-05-31)

# 1.28.0 (2022-05-24)

### Bug Fixes

- **deps:** update all dependencies j:cdx-227 ([#790](https://github.com/coveo/cli/issues/790)) ([2843a12](https://github.com/coveo/cli/commits/2843a125bd27f19eaac4dba7171d55a1e2d250cd))

## 1.27.1 (2022-04-20)

### Bug Fixes

- **deps:** update all dependencies j:cdx-227 ([#757](https://github.com/coveo/cli/issues/757)) ([0277354](https://github.com/coveo/cli/commits/027735409555eafa1c208bcd8aef02a36cd22847))
- **deps:** update all dependencies j:cdx-227 ([#758](https://github.com/coveo/cli/issues/758)) ([e381804](https://github.com/coveo/cli/commits/e381804d509cef83a3ec6ef7a39995a4d5c38bd4))

# 1.27.0 (2022-04-14)

# 1.26.0 (2022-03-31)

### Bug Fixes

- **deps:** update all dependencies j:cdx-227 ([#666](https://github.com/coveo/cli/issues/666)) ([09ea2e7](https://github.com/coveo/cli/commits/09ea2e7fd0bac33f2c2c1e91fa5d77adca159aa0))
- **deps:** update all dependencies j:cdx-227 ([#723](https://github.com/coveo/cli/issues/723)) ([88a18fd](https://github.com/coveo/cli/commits/88a18fdc3f035959a3ad13bb1aa30289a53bff72))

# 1.25.0 (2022-02-16)

### Bug Fixes

- **deps:** update all dependencies j:cdx-227 ([#623](https://github.com/coveo/cli/issues/623)) ([72d6fde](https://github.com/coveo/cli/commits/72d6fde0affc22b0c6263edbd6dcb8642e60f894))
- **search token server:** add express security plugins ([#624](https://github.com/coveo/cli/issues/624)) ([7b88321](https://github.com/coveo/cli/commits/7b88321a60d8f31aff287e6054e1944158c08627))

# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.24.1](https://github.com/coveo/cli/compare/v1.24.0...v1.24.1) (2022-01-18)

**Note:** Version bump only for package @coveo/search-token-server

# [1.24.0](https://github.com/coveo/cli/compare/v1.23.3...v1.24.0) (2022-01-18)

### Bug Fixes

- **deps:** update all dependencies j:cdx-227 ([#619](https://github.com/coveo/cli/issues/619)) ([8c739f8](https://github.com/coveo/cli/commit/8c739f870b94a6ba3d1adc847b4dea8f1bb4b05f))

## [1.23.3](https://github.com/coveo/cli/compare/v1.23.2...v1.23.3) (2022-01-14)

**Note:** Version bump only for package @coveo/search-token-server

## [1.23.2](https://github.com/coveo/cli/compare/v1.23.1...v1.23.2) (2022-01-13)

### Bug Fixes

- **deps:** update all dependencies j:cdx-227 ([#609](https://github.com/coveo/cli/issues/609)) ([f7d0f95](https://github.com/coveo/cli/commit/f7d0f95c7578b12f462b120b6f56d10d252a072d))

## [1.23.1](https://github.com/coveo/cli/compare/v1.23.0...v1.23.1) (2022-01-12)

**Note:** Version bump only for package @coveo/search-token-server

# [1.23.0](https://github.com/coveo/cli/compare/v1.22.2...v1.23.0) (2022-01-12)

### Bug Fixes

- **deps:** update all dependencies j:cdx-227 ([#605](https://github.com/coveo/cli/issues/605)) ([c7ffa16](https://github.com/coveo/cli/commit/c7ffa162b862dd2790506976b6a0c49bc9b7b5f6))

## [1.22.2](https://github.com/coveo/cli/compare/v1.22.1...v1.22.2) (2021-12-09)


src/scenario/scenario.ts
export function buildScenario() {
  async function run() {
    addItemToCart()
    
    openCart()
    await sleep(0.5)
    navigateToCheckout()

    pay()
    await sleep(0.5)
    
    reportResult()
    navigateToHome()
  }

  return {
    run
  }
}

function addItemToCart() {
  clickButton('.add-to-cart-btn')
}

function openCart() {
  clickButton('#cart-btn')
}

async function navigateToCheckout() {
  clickButton('#checkout-btn')
}

function pay() {
  clickButton('.pay-btn')
}

function navigateToHome() {
  clickButton('.home-btn')
}

function clickButton(selector: string) {
  const el = document.querySelector<HTMLButtonElement>(selector)
  el && el.click()
}

function sleep(seconds: number) {
  return new Promise(resolve => setTimeout(resolve, seconds * 1000));
}

function reportResult() {

}

src/Components/useCart.ts
import { Result } from "@coveo/headless";
import { buildStore } from "../common/store";

export interface CartItem {
  product: Result;
  quantity: number;
}

const store = buildStore<CartItem[]>([])

export function useCart() {
  const { set, subscribe } = store;

  function findCartItemIndex(product: Result) {
    return store.value.findIndex(item => item.product.uniqueId === product.uniqueId)
  }

  function addProduct(product: Result) {
    const index = findCartItemIndex(product)
    index === -1 ? addNewItem(product) : incrementItem(product)
  }

  function addNewItem(product: Result) {
    const newItems = [...store.value, { product, quantity: 1 }]
    set(newItems)
  }

  function incrementItem(product: Result) {
    const newItems = store.value.map(item => {
      const found = item.product.uniqueId === product.uniqueId
      return found ? ({ ...item, quantity: item.quantity + 1 }) : item;
    })

    set(newItems)
  }

  function removeProduct(product: Result) {
    const newItems = store.value.filter(item => item.product.uniqueId !== product.uniqueId)
    set(newItems)
  }

  function removeAll() {
    set([])
  }

  function get() {
    return store.value
  }

  return {
    addProduct,
    removeProduct,
    removeAll,
    subscribe,
    get
  }
}

scripts/port-allocator.mjs
import {
  appendFileSync,
  truncateSync,
  existsSync,
  writeFileSync,
} from "node:fs";
import { EOL } from "node:os";
import getPort, { makeRange } from "get-port";
import { config } from "dotenv";
config();

const preferedWebAppPort = 3000;
const portRangeFallback = [3000, 3999];

const ensureEnvironmentFile = () => {
  if (!existsSync(".env")) {
    writeFileSync(".env", "");
  }
};

const updateEnvFile = (applicationPort, serverPort) => {
  const env = {
    ...getDefaultEnvVariables(),
    PORT: applicationPort,
    REACT_APP_SERVER_PORT: serverPort,
  };

  truncateSync(".env");
  for (const [key, value] of Object.entries(env)) {
    appendFileSync(".env", `${key}=${value}${EOL}`);
  }
};

const getDefaultEnvVariables = () => {
  return {
    REACT_APP_PLATFORM_URL: process.env["REACT_APP_PLATFORM_URL"],
    REACT_APP_ORGANIZATION_ID: process.env["REACT_APP_ORGANIZATION_ID"],
    REACT_APP_API_KEY: process.env["REACT_APP_API_KEY"],
    REACT_APP_USER_EMAIL: process.env["REACT_APP_USER_EMAIL"],
    PORT: process.env["PORT"],
    REACT_APP_SERVER_PORT: process.env["REACT_APP_SERVER_PORT"],
  };
};

const allocatePorts = async () => {
  const applicationPort = await getNextAvailablePorts(
    process.env.PORT || preferedWebAppPort
  );
  const serverPort = await getNextAvailablePorts(
    process.env.REACT_APP_SERVER_PORT
  );
  updateEnvFile(applicationPort, serverPort);
};

const getNextAvailablePorts = async (preferedPort) => {
  if (typeof preferedPort !== "number") {
    preferedPort = isNaN(parseInt(preferedPort))
      ? null
      : parseInt(preferedPort);
  }

  if (await isPortAvailable(preferedPort)) {
    return preferedPort;
  }
  return await getPort({ port: makeRange(...portRangeFallback) });
};

const isPortAvailable = async (port) => {
  const availablePort = await getPort({ port });
  return availablePort === port;
};

const main = async () => {
  ensureEnvironmentFile();
  await allocatePorts();
};

main();


src/scenario/useCoveoAnalytics.ts
import 'coveo.analytics'
import { AnalyticsClientSendEventHook, handleOneAnalyticsEvent } from 'coveo.analytics';
import { useEventStore } from './useEventStore'

type CoveoUAGlobal = typeof handleOneAnalyticsEvent

declare global {
  var coveoua: CoveoUAGlobal;
  var coveoanalytics: any;
}

export function useCoveoAnalytics() {
  const { add } = useEventStore();
  const hook: AnalyticsClientSendEventHook = (type, payload) => {
    const event = { type, payload }
    add(event)
    return payload
  }

  const client = window.coveoanalytics?.SimpleAnalytics.coveoua.client

  if (client && !client.hookAttached) {
    window.coveoanalytics.SimpleAnalytics.coveoua.client.registerBeforeSendEventHook(hook);
    client.hookAttached = true;
  }

  const coveoua = window.coveoua || (() => console.log('noop coveoua'));
  return { coveoua }
}

src/scenario/useScenario.ts
import { CartItem, useCart } from "../Components/useCart";
import { AnalyticsEvent, useEventStore } from "./useEventStore";
import { EventReport, ReportItem, useScoreCardStore } from "./useScoreCard";

export function useScenario() {
  const { reset: resetEventStore, get: getEvents } = useEventStore()
  const { removeAll: emptyCart, get: getCart } = useCart();
  const { set: setScoreCard } = useScoreCardStore()

  async function run() {
    resetEventStore()
    emptyCart()
    addItemToCart()

    isCartClosed() && openCart()
    await sleep()
    navigateToCheckout()
    await sleep()
    pay()
    await sleep()
    navigateToHome()
    await sleep()

    const items = getCart()
    const events = getEvents()
    emptyCart()

    const scoreCard = analyze(items, events)
    setScoreCard(scoreCard)
  }

  return {
    run
  }
}

function addItemToCart() {
  clickButton('.add-to-cart-btn')
}

function isCartClosed() {
  return !document.querySelector('.cart-list')
}

function openCart() {
  clickButton('#cart-btn')
}

async function navigateToCheckout() {
  clickButton('#checkout-btn')
}

function pay() {
  clickButton('.pay-btn')
}

function navigateToHome() {
  clickButton('.home-btn')
}

function clickButton(selector: string) {
  const el = document.querySelector<HTMLButtonElement>(selector)
  el && el.click()
}

function sleep(seconds: number = 0.5) {
  return new Promise(resolve => setTimeout(resolve, seconds * 1000));
}

function analyze(items: CartItem[], events: AnalyticsEvent[]) {
  return [
    checkAddToCart(items[0], events[0], 0),
    checkCheckoutPageView(events[1]),
    checkPurchase(events[2], items),
    checkHomePageView(events[3])
  ]
}

type LoggedAnalyticsEvent = AnalyticsEvent | undefined

function checkAddToCart(cartItem: CartItem, event: LoggedAnalyticsEvent, index: number): EventReport {
  return {
    event: 'addToCart',
    payload: getPayload(event),
    report: event ? getAddToCartEventReport(cartItem, event, index) : [],
    missing: !event
  }
}

function getAddToCartEventReport(cartItem: CartItem, event: AnalyticsEvent, index: number) {
  return [
    assertPayload(event, 'action', 'add'),
    ...assertProduct(event, cartItem, index)
  ]
}

function checkCheckoutPageView(event: LoggedAnalyticsEvent): EventReport {
  return {
    event: 'checkout pageview',
    payload: getPayload(event),
    report: event ? [
      assertPayload(event, 'hitType', 'pageview'),
      assertPayload(event, 'page', '/checkout')
    ] : [],
    missing: !event
  }
}

function getPayload(event: LoggedAnalyticsEvent) {
  return event ? event.payload : {}
}

function checkHomePageView(event: LoggedAnalyticsEvent): EventReport {
  return {
    event: 'search pageview',
    payload: getPayload(event),
    report: event ? [
      assertPayload(event, 'hitType', 'pageview'),
      assertPayload(event, 'page', '/search')
    ] : [],
    missing: !event
  }
}

function checkPurchase(event: LoggedAnalyticsEvent, items: CartItem[]): EventReport {
  return {
    event: 'purchase',
    payload: getPayload(event),
    report: event ? [
      assertPayload(event, 'action', 'purchase'),
      ...items.flatMap((item, i) => assertProduct(event, item, i)),
      assertPayload(event, 'id', '931cbf0c-07b0-4be1-91bb-448b3d82addc-1677170972912'),
      assertPayload(event, 'revenue', sumCart(items)),
    ] : [],
    missing: !event
  }
}

function sumCart(items: CartItem[]) {
  return items
    .map(item => {
      const rawPrice = item.product.raw.price
      const price = typeof rawPrice === 'number' ? rawPrice : 0;
      return price * item.quantity
    })
    .reduce((acc, curr) => acc + curr, 0)
}

function assertProduct(event: AnalyticsEvent, cartItem: CartItem, index: number): ReportItem[] {
  const prefix = `pr${index + 1}`
  const id = `${prefix}id`
  const category = `${prefix}ca`
  const price = `${prefix}pr`
  const name = `${prefix}nm`
  const quantity = `${prefix}qt`

  return [
    assertPayload(event, id, cartItem.product.clickUri),
    assertPayload(event, category, cartItem.product.clickUri),
    assertPayload(event, price, cartItem.product.clickUri),
    assertPayload(event, name, cartItem.product.clickUri),
    assertPayload(event, quantity, cartItem.quantity),
  ]
}

function assertPayload(event: AnalyticsEvent, key: string, value: any): ReportItem {
  if (event.payload[key] !== value) {
    const message = `expected ${key} to be "${value}"`
    return { valid: false, message }
  }

  return { valid: true, message: `${key} is correct` }
}
